src/Serialization/main.jl

```diff
@@ -73,36 +92,117 @@ function decodeType(input::String)
+function has_elem_basic_encoding(obj::T) where T <: Ring
+    return is_basic_serialization_type(elem_type(obj))
+end
+
+has_basic_encoding(obj::T) where T = is_basic_serialization_type(T)
+
+function has_basic_encoding(obj::T) where T <: RingElem
+    return has_elem_basic_encoding(parent(obj))
+end
+
+has_elem_basic_encoding(obj::T) where T = false
+has_elem_basic_encoding(obj::ArbField) = true
+has_elem_basic_encoding(obj::AcbField) = true
+has_elem_basic_encoding(obj::FqField) = absolute_degree(obj) == 1
+has_elem_basic_encoding(obj::Nemo.fpField) = true
+has_elem_basic_encoding(obj::Nemo.FpField) = true
+has_elem_basic_encoding(obj::Nemo.zzModRing) = true
+has_elem_basic_encoding(obj::FlintPadicField) = true
+has_elem_basic_encoding(obj::TropicalSemiring) = true
 
 
 ################################################################################
 # High level
-function save_type_dispatch(s::SerializerState, obj::T) where T
-    if is_basic_serialization_type(T) && s.depth != 0
-        return save_internal(s, obj)
+
+function load_ref(s::DeserializerState, dict::Dict)
+    if !haskey(dict, :id)
+        return load_unknown_type(s, dict)
     end
 
-    if !isprimitivetype(T) && !Base.issingletontype(T)
-        # if obj is already serialzed, just output
-        # a backref
-        ref = get(s.objmap, obj, nothing)
-        if ref !== nothing
-            return Dict{Symbol, Any}(
-              :type => backref_sym,
-              :id => string(ref),
-              :version => 1, # ???
-              )
-        end
-        # otherwise,
-        ref = s.objmap[obj] = uuid4()
+    id = dict[:id]
+
+    if haskey(s.objs, UUID(id))
+        loaded_ref = s.objs[UUID(id)]
     else
-        ref = nothing
+        ref_dict = s.refs[Symbol(id)]
+        ref_dict[:id] = id
+        loaded_ref = load_unknown_type(s, ref_dict)
     end
+    return loaded_ref
+end
 
-    result = Dict{Symbol, Any}(:type => encodeType(T))
+function save_as_ref(s::SerializerState, obj::T) where T
+    if has_elem_basic_encoding(obj) && s.depth != 0
+        return save_type_dispatch(s, obj)
+    end
+
+    # find ref or create one
+    ref = get(s.objmap, obj, nothing)
     if ref !== nothing
-        result[:id] = string(ref)
+        return Dict{Symbol, Any}(
+            :type => backref_sym,
+            :id => string(ref),
+        )
+    end
+
+    ref = s.objmap[obj] = uuid4()
+    result = Dict{Symbol, Any}(:type => encodeType(T))
+
+    if Base.issingletontype(T)
+        return result
     end
+
+    # invoke the actual serializer
+    result[:data] = save_internal(s, obj)
+    s.refs[Symbol(ref)] = result
+
+    return Dict{Symbol, Any}(
+        :type => backref_sym,
+        :id => string(ref),
+    )
+end
+
+function save_type_dispatch(s::SerializerState, obj::T) where T
+    # this is used when serializing basic types like "3//4"
+    # when s.depth == 0 file should know it belongs to QQ
+    if s.depth != 0 && has_basic_encoding(obj)
```

Comment:
So depth 0 is somewhat special. Maybe we need some test cases just for it, to make sure things work right?